torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -324,127 +258,141 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
+
+        # - go through the per-shard states, which are all indexed locally
+        for rank, s in enumerate(self._all_states):
+            # -- match the local indexing and the global partition, update the corresponding saved state globally
+            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
+                local_index_to_param_id = {
+                    i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
+                }
+
+                for local_param_index in local_pg["params"]:
+                    # Update the state, if any
+                    if local_param_index in s["state"].keys():
+                        global_id = self.param_to_index[local_index_to_param_id[local_param_index]]
+                        state_dict["state"][global_id] = s["state"][local_param_index]
+
+        return state_dict
 
-        # Force a re-partitioning, in case the model changed with the new state
-        self._partition_parameters.clear()
-        self._per_device_params.clear()
-        self._param_rank.clear()
+    def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        """Update the consolidated state_dict list, one per rank.
 
-        # Update the bucketing strategy accordingly
-        self._setup_bucket_strategy()
+        .. warning: This needs to be called on all replicas"""
 
-    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
-        """Restore the global parameter groups as well as the shard.
+        # Sync lr and other attributes in case its been updated
+        ZeroRedundancyOptimizer._sync_param_groups(self.param_groups, self.optim.param_groups)
 
-        Arguments:
-            state_dict (dict): optimizer state. Should be an object returned
-                from a call to :meth:`state_dict`
-        """
+        # Pull the sharded state from all the other replicas
+        # Store all the states in order, rank by rank
 
-        # Check whether we got a local or global dict
-        if "local_state_dict" in state_dict and state_dict["local_state_dict"]:
-            self.load_local_state_dict(state_dict)
-        else:
-            # Dispatch this rank's state dictionary to the wrapped shard optimizer
-            self.load_local_state_dict(ZeroRedundancyOptimizer.rank_local_state_dict(self.rank, state_dict))
+        # NOTE: In practice, `broadcast` is used, which is wasteful (gather would have been appropriate)
+        # compatibility issues with some backends make the use of broadcast mandatory for now.
+        # a possible follow up would be to move all sharded state management to RPC RRef
 
-    def local_state_dict(self) -> Dict:
-        """Gets this rank's ``state_dict``.
+        self._all_states = []
 
-        Returns:
-            The state of the optimizer as a :class:`dict`.
-            It contains two entries:
+        with torch.cuda.device(self._device) if self._device.type == torch.device(
```

Comment:
not  changed, except for the move to pytorch's `broadcast_object_list`. Getting the device right for all platforms is not completely obvious I think